File: bot/rate_limiter.py
"""Rate limiting for Telegram API"""

import asyncio
from config import CONFIG


class RateLimiter:
    def __init__(self):
        self.global_last_send = 0.0
        self.group_last_send: dict[int, float] = {}
        self.send_lock = asyncio.Lock()
        self.user_locks: dict[int, asyncio.Lock] = {}
        self.active_users: set[int] = set()
    
    async def safe_send(self, chat_id: int, coro):
        """Send with rate limiting"""
        async with self.send_lock:
            now = asyncio.get_event_loop().time()
            
            # Global rate limit
            global_wait = CONFIG.global_min_interval - (now - self.global_last_send)
            if global_wait > 0:
                await asyncio.sleep(global_wait)
            
            # Group rate limit
            if chat_id < 0:
                last_group = self.group_last_send.get(chat_id, 0)
                group_wait = CONFIG.group_min_interval - (now - last_group)
                if group_wait > 0:
                    await asyncio.sleep(group_wait)
                self.group_last_send[chat_id] = asyncio.get_event_loop().time()
            
            self.global_last_send = asyncio.get_event_loop().time()
            
            for attempt in range(CONFIG.max_retries):
                try:
                    return await coro
                except Exception as e:
                    error_str = str(e)
                    if "429" in error_str or "retry_after" in error_str.lower():
                        wait = 30 + 5  # default retry + buffer
                        print(f"[rate-limit] 429, waiting {wait}s ({attempt+1}/{CONFIG.max_retries})")
                        if attempt < CONFIG.max_retries - 1:
                            await asyncio.sleep(wait)
                    else:
                        print(f"[send] Error: {error_str[:100]}")
                        return None
            return None
    
    def get_user_lock(self, user_id: int) -> asyncio.Lock:
        if user_id not in self.user_locks:
            self.user_locks[user_id] = asyncio.Lock()
        return self.user_locks[user_id]
    
    def can_accept_user(self, user_id: int) -> bool:
        if user_id in self.active_users:
            return True
        return len(self.active_users) < CONFIG.max_concurrent
    
    def mark_active(self, user_id: int):
        self.active_users.add(user_id)
    
    def mark_inactive(self, user_id: int):
        self.active_users.discard(user_id)


# Singleton instance
rate_limiter = RateLimiter()

File: bot/src/rate-limiter.ts
/**
 * Rate limiting and user locking
 */

import { CONFIG } from './config.js';

// Global rate limiter
let globalLastSend = 0;
const lastGroupMessage = new Map<number, number>();
let sendMutex = Promise.resolve();

// Safe send with global rate limiting
export async function safeSend<T>(
  chatId: number,
  fn: () => Promise<T>,
  maxRetries = CONFIG.rateLimit.maxRetries
): Promise<T | null> {
  const myTurn = sendMutex;
  let release: () => void;
  sendMutex = new Promise(r => { release = r; });
  
  await myTurn;
  
  try {
    const now = Date.now();
    const globalWait = CONFIG.rateLimit.globalMinInterval - (now - globalLastSend);
    if (globalWait > 0) {
      await new Promise(r => setTimeout(r, globalWait));
    }
    
    if (chatId < 0) {
      const lastGroup = lastGroupMessage.get(chatId) || 0;
      const groupWait = CONFIG.rateLimit.groupMinInterval - (Date.now() - lastGroup);
      if (groupWait > 0) {
        await new Promise(r => setTimeout(r, groupWait));
      }
      lastGroupMessage.set(chatId, Date.now());
    }
    
    globalLastSend = Date.now();
    
    for (let attempt = 1; attempt <= maxRetries; attempt++) {
      try {
        return await fn();
      } catch (e: any) {
        if (e.response?.error_code === 429) {
          const retryAfter = (e.response?.parameters?.retry_after || 30) + CONFIG.rateLimit.retryBuffer;
          console.log(`[rate-limit] 429, waiting ${retryAfter}s (${attempt}/${maxRetries})`);
          if (attempt < maxRetries) {
            await new Promise(r => setTimeout(r, retryAfter * 1000));
            globalLastSend = Date.now();
          }
        } else {
          console.error(`[send] Error: ${e.message?.slice(0, 100)}`);
          return null;
        }
      }
    }
    return null;
  } finally {
    release!();
  }
}

// Per-user rate limiter
const userLocks = new Map<number, Promise<void>>();

export async function withUserLock<T>(userId: number, fn: () => Promise<T>): Promise<T> {
  const existing = userLocks.get(userId);
  let resolve: () => void;
  const myLock = new Promise<void>(r => { resolve = r; });
  userLocks.set(userId, myLock);
  
  if (existing) await existing;
  
  try {
    return await fn();
  } finally {
    resolve!();
    if (userLocks.get(userId) === myLock) {
      userLocks.delete(userId);
    }
  }
}

// Concurrent users limiter
const activeUsers = new Set<number>();
let maxConcurrentUsers = CONFIG.users.maxConcurrent;

export function setMaxConcurrentUsers(max: number) {
  maxConcurrentUsers = max;
}

export function canAcceptUser(userId: number): boolean {
  if (activeUsers.has(userId)) return true;
  return activeUsers.size < maxConcurrentUsers;
}

export function markUserActive(userId: number) {
  activeUsers.add(userId);
}

export function markUserInactive(userId: number) {
  activeUsers.delete(userId);
}
